src/agent/firebasecontroller.ts

```diff
@@ -0,0 +1,224 @@
+
+    // Calculate the debuggee id as the hash of the object.
+    // This MUST be consistent across all debuggee instances.
+    this.debuggeeId = crypto
+      .createHash('md5')
+      .update(JSON.stringify(debuggee))
+      .digest('hex');
+    debuggee.id = this.debuggeeId;
+
+    const debuggeeRef = this.db.ref(`cdbg/debuggees/${this.debuggeeId}`);
+    debuggeeRef.set(debuggee);
+
+    // TODO: Handle errors.  I can .set(data, (error) => if (error) {})
+    const agentId = 'unsupported';
+    callback(null, {debuggee, agentId});
+  }
+
+  /**
+   * Update the server about breakpoint state
+   * @param {!Debuggee} debuggee
+   * @param {!Breakpoint} breakpoint
+   * @param {!Function} callback accepting (err, body)
+   */
+  updateBreakpoint(
+    debuggee: Debuggee,
+    breakpoint: stackdriver.Breakpoint,
+    callback: (err?: Error, body?: {}) => void
+  ): void {
+    console.log('updating a breakpoint');
+    assert(debuggee.id, 'should have a registered debuggee');
+
+    // By default if action is not present, it's a snapshot, so for it to be
+    // a logpoint it must be present and equal to 'LOG', anything else is a
+    // snapshot.
+    const is_logpoint = breakpoint.action === 'LOG';
+    const is_snapshot = !is_logpoint;
+
+    if (is_snapshot) {
+      breakpoint.action = 'CAPTURE';
+    }
+
+    breakpoint.isFinalState = true;
+
+    const breakpoint_map: {[key: string]: any} = {...breakpoint};
+
+    // Magic value. Firebase RTDB will replace the {'.sv': 'timestamp'} with
+    // the unix time since epoch in milliseconds.
+    // https://firebase.google.com/docs/reference/rest/database#section-server-values
+    breakpoint_map['finalTimeUnixMsec'] = {'.sv': 'timestamp'};
+
+    // TODO: error handling from here on
+    this.db
+      .ref(`cdbg/breakpoints/${this.debuggeeId}/active/${breakpoint.id}`)
+      .remove();
+
+    if (is_snapshot) {
+      // We could also restrict this to only write to this node if it wasn't
+      // an error and there is actual snapshot data. For now though we'll
+      // write it regardless, makes sense if you want to get everything for
+      // a snapshot it's at this location, regardless of what it contains.
+      this.db
+        .ref(`cdbg/breakpoints/${this.debuggeeId}/snapshot/${breakpoint.id}`)
+        .set(breakpoint_map);
+      // Now strip the snapshot data for the write to 'final' path.
+      const fields_to_strip = [
+        'evaluatedExpressions',
+        'stackFrames',
+        'variableTable',
+      ];
+      fields_to_strip.forEach(field => delete breakpoint_map[field]);
+    }
+
+    this.db
+      .ref(`cdbg/breakpoints/${this.debuggeeId}/final/${breakpoint.id}`)
+      .set(breakpoint_map);
+
+    // Indicate success to the caller.
+    callback();
+  }
+
+  subscribeToBreakpoints(
+    debuggee: Debuggee,
+    callback: (err: Error | null, breakpoints: stackdriver.Breakpoint[]) => void
+  ): void {
+    console.log('Started subscription for breakpoint updates');
```

Comment:
I would use https://nodejs.org/api/util.html#utildebuglogsection-callback rather than `console.log`, that way you don't have to remember to remove all of the debug lines from your code, and if you make the debug lines descriptive, it's even worth leaving them in.

Reply:
Done.  Thanks.